web/client-api/src/main/java/io/deephaven/web/client/api/grpc/MultiplexedWebsocketTransport.java

```diff
@@ -121,9 +121,77 @@ public void sendFallback(Transport transport) {
 
-    private final WebSocket webSocket;
+    static class ActiveTransport {
+        private static final Map<String, ActiveTransport> activeSockets = new HashMap<>();
+        private final WebSocket webSocket;
+        private boolean closing;
+        private int activeCount = 0;
+
+        /**
+         * Gets a websocket transport for the grpc url.
+         *
+         * @param grpcUrl The URL to access - the full path will be used (with ws/s instead of http/s) to connect to the
+         *        service, but only the protocol+host+port will be used to share the instance.
+         * @return a websocket instance to use to connect, newly created if necessary
+         */
+        public static ActiveTransport get(String grpcUrl) {
+            URL urlWrapper = new URL(grpcUrl);
+            if (urlWrapper.protocol.equals("http:")) {
+                urlWrapper.protocol = "ws:";
+            } else {
+                urlWrapper.protocol = "wss:";
+            }
+            String actualUrl = urlWrapper.toString();
+            urlWrapper.pathname = "/";
+            String key = urlWrapper.toString();
+            return activeSockets.computeIfAbsent(key, url -> new ActiveTransport(key, actualUrl));
+        }
+
+        /**
+         * @param key URL to use as the key entry to reuse the transport
+         * @param actualUrl the url to connect to
+         */
+        private ActiveTransport(String key, String actualUrl) {
+            this.webSocket = new WebSocket(actualUrl, new String[] {MULTIPLEX_PROTOCOL, SOCKET_PER_STREAM_PROTOCOL});
+
+            webSocket.binaryType = "arraybuffer";
+
+            webSocket.addEventListener("message", event -> {
+                MessageEvent<ArrayBuffer> messageEvent = Js.uncheckedCast(event);
+                // read the message, check if it was a GO_AWAY
+                int streamId = new DataView(messageEvent.data, 0, 4).getInt32(0);
+                if (streamId == Integer.MAX_VALUE) {
+                    // Server sent equiv of H2 GO_AWAY, time to wrap up.
+                    // ACK the message
+                    new WebsocketFinishSignal().send(webSocket, Integer.MAX_VALUE);
+
+                    // We can attempt to create new transport instances, but cannot use this one any longer for new
+                    // streams (and any new one is likely to fail unless some new server is ready for us)
+                    activeSockets.remove(key);
+
+                    // Mark that this transport should be closed when existing streams finish
+                    this.closing = true;
+                    if (activeCount == 0) {
+                        webSocket.close();
+                    }
+                }
+            });
+        }
+
+        private void retain() {
+            activeCount++;
+        }
+
+        private void release() {
+            activeCount--;
+            if (activeCount == 0 && closing) {
+                webSocket.close();
+            }
+        }
```

Comment:
Are there any threading concerns we have to worry about? I'm assuming this is a different sync model wrt JS?

Reply:
No threads on JS, correct, various java keywords have no effect when transpiled to js, as they have no js analog.